File: python/src/algorithm/geometry/solids/rounds/sphere.py
import math

from algorithm.geometry import SolidFigure
from algorithm.utils import validation


class Sphere(SolidFigure):
    """
    Sphere
    ======

    A perfectly round 3D shape where every point on the surface is equidistant
    from the center (e.g., a basketball). It has no edges or vertices.

    Example
    -------
    ```
    from algorithm.geometry.solids.rounds import Sphere

    sphere: Sphere = Sphere(5)

    print(f"Sphere Surface Area: {sphere.surface_area()}")
    print(f"Sphere Volume: {sphere.volume()}")
    ```
    """

    def __init__(self, radius: float):
        validation.validate_positive(radius, "radius")
        self._radius = radius

    @property
    def radius(self) -> float:
        """
        Returns the radius of the sphere.
        """
        return self._radius

    @property
    def diameter(self) -> float:
        """
        Returns the diameter of the sphere.
        """
        return self._radius * 2

    def surface_area(self) -> float:
        return 4 * math.pi * pow(self._radius, 2)

    def volume(self) -> float:
        return (4 / 3) * math.pi * pow(self._radius, 3)

File: java/src/main/java/algorithm/geometry/solids/Sphere.java
package algorithm.geometry.solids;

import algorithm.geometry.Solid;
import algorithm.utils.Validation;

public final class Sphere implements Solid {
    private double radius;

    public Sphere(double radius) throws Exception {
        Validation.validatePositive(radius, "radius");
        this.radius = radius;
    }

    public double getRadius() {
        return radius;
    }

    @Override
    public double volume() {
        return (4.0 / 3.0) * Math.PI * Math.pow(radius, 3);
    }

    @Override
    public double surfaceArea() {
        return 4 * Math.PI * Math.pow(radius, 2);
    }
}
